### DISCUSSION TITLE
Expanding Collection Elements in Queries

### DISCUSSION THREAD

**Discussion by n1njab0b** [datalust/seq-tickets#1833] [Help]
If I am logging events like this:
`Log.Information("Updated {ProductId} with categories {Categories}", "product-32", new[] { "drinks", "coffees" });`

How can I query the events so that the categories "drinks" and "coffees" show up in separate records?

**Comment by nblumhardt** [comment]
Hi @n1njab0b,

Seq's underlying query engine (Flare) supports this functionality through its `unnest()` set function - following PostgreSQL's `cross join lateral unnest()` feature.

Unfortunately, it's not exposed in Seq's user-facing query syntax - yet? - because we're unsure exactly how the syntax should look given the simplifications used there.

I have a very incomplete and shaky spike where the `cross join lateral` is implicit and you might write:

```sql
select min(ProductId)
from stream
group by unnest(Categories)
```

and get what you'd expect (lowest product id by category), or:

```sql
select ProductId, unnest(Categories)
from stream
limit 5
```

to get a flat list of product id/category pairs, but a huge number of questions remain to be answered.

It's something we're all interested in figuring out, but it's not clear when we'll be able to assemble a good proposal for a watertight version of the feature... 🤔 

In the meantime, `distinct(Categories)` and `count(distinct(Categories))` both work the way you're looking for, in case either of those partially covers your use case.

**Comment by n1njab0b** [comment]
Thanks so much for the detailed response @nblumhardt,

I had some luck with distinct, but I am having trouble applying the unnest suggestions you provided, and I think it's because I understated the complexity of my challenge in my original post.

I said I was doing this,
`Log.Information("Updated {ProductId} with categories {Categories}", "product-32", new[] { "drinks", "coffees" });`

When a more accurate example would be this:
```
Log.Information(
 "Updated {ProductId} with categories {Categories}", 
 "product-32", 
 new Category[] { 
    new Category(CategoryId:1,Label:"drinks"),
    new Category(CategoryId:2,Label:"coffees")
  });
```

What I am trying to fetch from Seq is 2 records with ProductId,CategoryId, and Label.

When I attempt to use the unnest examples you provided, I receive the error: `Invalid query: set functions must be declared in a 'from ... lateral' clause`.

Is what I am trying to do possible with the current engine?  If not, is there a different way that I can "shape" my log events to make it more conducive to query/export?  I could certainly log them as separate events, but that seems antithetical to the benefits of structured logging.

P.S. In my search for details on the Flare engine, I read the [Tour of Seq's Storage Engine](https://blog.datalust.co/a-tour-of-seqs-storage-engine/).  It was fascinating and it also gave me an appreciation for all the challenges that you face trying to store and fetch this data in the most efficient, reliable, and performant way possible.  Just want you to know your efforts are appreciated.

**Reply by nblumhardt** [reply]
That's awesome, thanks for the follow-up and feedback :-)

No, `unnest()` isn't supported today (as you noticed, it's accepted syntax but the back-end rejects it because `from lateral` isn't available; we should probably tighten up the error reporting here).

It just occurred to me that you might get what you're after using:

```sql
select distinct(Categories)
from stream
group by ProductId
```

The result won't be in full rowset form (both the category id and label will be JSON in the same column) but the data should be pretty much what you're looking for. Let me know if this helps 🤞

**Comment by n1njab0b** [comment]
So close!

Indeed, using distinct provides me with 1 JSON element per record, which I could definitely work worth, except that I can't download it. :/

When I hit the "Export as CSV" button, nothing happens.  When I look behind the scenes via Chrome Developer Tools, I see that the /api/deferred endpoint is responding with a status 400 (bad request) error, `{"Error":"Computation timed out or was interrupted."}` which is odd because the error is immediate. This is on Seq 2023.1.9101.

In the meantime, I can get by with a copy/paste of the search results, it's just not super optimal.  If I get really desperate, I suppose I could script an API client to download the log records and parse them.

**Comment by liammclennan** [comment]
Hi @n1njab0b 

That failure of CSV export was intermittent, so you might be able to succeed by trying a few times, especially when the server is not busy. If you are able to update to the latest Seq it has been fixed.

**Reply by nblumhardt** [reply]
(Tracking issue for the CSV download failure - https://github.com/datalust/seq-tickets/issues/1810.)

**Reply by n1njab0b** [reply]
Thanks @liammclennan,

I gave it 10 tries and unfortunately it never succeeded.  I'll try again after we update to the latest version, which will be during our next maintenance window.

**Reply by n1njab0b** [reply]
Good news, after updating to the latest version of Seq, I have confirmed that the CSV download functionality is fixed.

Thanks again for the help guys.
